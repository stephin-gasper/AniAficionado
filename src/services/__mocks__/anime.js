import {
  INITIAL_LATEST_SUBBED_EPISODES_RESPONSE,
  LATEST_SUBBED_EPISODES_RESPONSE,
} from 'services/anime.mock';

export const getInitialLatestSubbedEpisodes = jest.fn(() =>
  Promise.resolve(INITIAL_LATEST_SUBBED_EPISODES_RESPONSE),
);

export const getLatestSubbedEpisodes = jest.fn(() =>
  Promise.resolve(LATEST_SUBBED_EPISODES_RESPONSE),
);
