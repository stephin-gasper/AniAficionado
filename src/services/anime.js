import * as cheerio from 'cheerio';
import {fetchHomepageHtml, fetchLatestSubbedEpisodes} from 'api/animixplay';
import formatDateToTimeAgo from 'utils/formatDateToTimeAgo';

let lastSubbedEpisodeDateAndTime;

const getEpisodeNumbers = episodeText => {
  const episodeNumberRegex = /EP (?<number>\d+)\/(?<total>\?|\d+)/;
  const result = episodeText.match(episodeNumberRegex).groups;
  return {
    latestEpisodeNumber: result.number,
    totalEpisodes: result.total,
  };
};

const getReleaseTimeText = dateText => formatDateToTimeAgo(new Date(dateText));

const getRatingText = text => {
  if (Number(text) === 0) return '?';
  const result = text.match(/(?<first>\d)(?<rest>\d+)/).groups;
  return `${result.first}.${result.rest}`;
};

export const getInitialLatestSubbedEpisodes = async () => {
  let list = [];
  try {
    const htmlString = await fetchHomepageHtml();
    if (htmlString) {
      const $ = cheerio.load(htmlString, {
        _useHtmlParser2: true,
      });

      $('#resultplace > .searchresult > li').each((i, el) => {
        lastSubbedEpisodeDateAndTime = $(el).find('.timetext').text();
        list[i] = {
          id: $(el).find('a').attr('href'),
          title: $(el).find('a').attr('title'),
          imageUrl: $(el).find('.resultimg').attr('src'),
          ...getEpisodeNumbers($(el).find('.infotext').text()),
          releaseTime: getReleaseTimeText(lastSubbedEpisodeDateAndTime),
          rating: getRatingText($(el).find('.rating').text()),
        };
      });
    }
  } catch {
    list = [];
  }
  return list;
};

export const getLatestSubbedEpisodes = async () => {
  const response = await fetchLatestSubbedEpisodes({
    episodeReleaseDateAndTime: lastSubbedEpisodeDateAndTime,
  });

  if (response) {
    lastSubbedEpisodeDateAndTime = response.last;
    return {
      list: response.result.map(item => ({
        id: item.url,
        title: item.title,
        imageUrl: item.picture,
        ...getEpisodeNumbers(item.infotext),
        releaseTime: getReleaseTimeText(item.timetop),
        rating: getRatingText(`${item.score}`),
      })),
      canLoadMoreResults: response.more,
    };
  }

  throw new Error('No response');
};
