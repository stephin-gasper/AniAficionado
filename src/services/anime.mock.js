export const INITIAL_LATEST_SUBBED_EPISODES_RESPONSE = [
  {
    id: '/v1/boruto-naruto-next-generations/ep267',
    imageUrl:
      'https://cdn.animixplay.to/i/2b749c9e3e272f2cf4b4f019ffe6c252.jpg',
    latestEpisodeNumber: '267',
    releaseTime: '14 hours ago',
    title: 'Boruto: Naruto Next Generations',
    totalEpisodes: '?',
    rating: '5.78',
  },
  {
    id: '/v1/one-piece/ep1033',
    imageUrl:
      'https://cdn.animixplay.to/i/73059c39fee46b7c378a3aa0bdd9b6a2.jpg',
    latestEpisodeNumber: '1033',
    releaseTime: '21 hours ago',
    title: 'One Piece',
    totalEpisodes: '?',
    rating: '8.67',
  },
  {
    id: '/v1/kakkou-no-iinazuke/ep21',
    imageUrl:
      'https://cdn.animixplay.to/i/41adf097b26cf486c64da6ada76adac2.jpg',
    latestEpisodeNumber: '21',
    releaseTime: '1 day ago',
    title: 'A Couple of Cuckoos',
    totalEpisodes: '24',
    rating: '?',
  },
];

export const LATEST_SUBBED_EPISODES_RESPONSE = {
  list: [
    {
      id: '/v1/poputepipikku-2nd-season/ep2',
      imageUrl: 'https://cdn.animixplay.to/min/mal/1237/128067.jpg',
      latestEpisodeNumber: '2',
      rating: '7.25',
      releaseTime: '1 month ago',
      title: 'Pop Team Epic Season 2',
      totalEpisodes: '12',
    },
    {
      id: '/v1/berserk-ougon-jidai-hen-memorial-edition/ep2',
      imageUrl:
        'https://cdn.animixplay.to/i/1321b74296956451a9be2f8eb05aaa10.jpg',
      latestEpisodeNumber: '2',
      rating: '7.62',
      releaseTime: '1 month ago',
      title: 'Berserk: The Golden Age Arc - Memorial Edition',
      totalEpisodes: '?',
    },
    {
      id: '/v1/ani-ni-tsukeru-kusuri-wa-nai-5/ep11',
      imageUrl:
        'https://cdn.animixplay.to/i/b138226a737f622fdd885b8066cb0123.jpg',
      latestEpisodeNumber: '11',
      rating: '7.07',
      releaseTime: '1 month ago',
      title: 'Ani ni Tsukeru Kusuri wa Nai! 5',
      totalEpisodes: '12',
    },
  ],
  canLoadMoreResults: true,
};
