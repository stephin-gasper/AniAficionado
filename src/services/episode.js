import * as cheerio from 'cheerio';
import {btoa} from 'react-native-quick-base64';
import {fetchEpisodeHtml, getAnimixplayBaseUrl} from 'api/animixplay';

const processFillerAndUpdateEpisodeList = (
  fillerListRegexMatchResult,
  episodesList,
) => {
  const episodesListCopy = [...episodesList];
  if (fillerListRegexMatchResult) {
    let parseSafeArrayString = `[${fillerListRegexMatchResult}]`;
    parseSafeArrayString = parseSafeArrayString.replace(/,]$/, ']'); // remove trailing comma
    parseSafeArrayString = parseSafeArrayString.replace(/'/g, '"'); // replace single quotes to double quotes
    const fillerList = JSON.parse(parseSafeArrayString);
    for (let outerIndex = 0; outerIndex < fillerList.length; outerIndex++) {
      const fillerData = fillerList[outerIndex];
      if (typeof fillerData === 'string' && fillerData.includes('-')) {
        const parseFillerData = fillerData.split('-', 2);
        const startIndex = parseInt(parseFillerData[0], 10);
        const endIndex = parseInt(parseFillerData[1], 10) + 1;
        for (
          let innerLoopIndex = startIndex;
          innerLoopIndex < endIndex;
          innerLoopIndex++
        ) {
          episodesListCopy[innerLoopIndex - 1] = {isFiller: true};
        }
      } else if (typeof fillerData === 'number') {
        episodesListCopy[fillerData - 1] = {isFiller: true};
      }
    }
  }
  return episodesListCopy;
};

class Episode {
  #streamingLinksObj;

  #gohostname = 'gogoplay.io';

  async getDetails(path) {
    try {
      const htmlString = await fetchEpisodeHtml(path);
      if (htmlString) {
        const $ = cheerio.load(htmlString, {
          _useHtmlParser2: true,
        });

        this.#streamingLinksObj = JSON.parse($('#epslistplace').text());
        let episodesList = new Array(this.#streamingLinksObj.eptotal).fill({
          isFiller: false,
        });
        let malid;
        // Use pattern search instead of eval to get variables because of dynamic fetch from third party
        $('script')
          .not((i, el) => $(el).attr('src'))
          .each((i, el) => {
            const scriptText = $(el).text().replace(/\s/g, '');
            const malidRegexMatchResult = scriptText.match(
              /malid='(?<result>\d+)/,
            )?.groups?.result;
            if (malidRegexMatchResult) {
              malid = malidRegexMatchResult;
            }
            const gohostnameRegexMatchResult = scriptText.match(
              /copt='#gh=(?<result>\S+)';\//,
            )?.groups?.result;
            if (gohostnameRegexMatchResult) {
              this.#gohostname = gohostnameRegexMatchResult;
            }
            const fillerListRegexMatchResult = scriptText.match(
              /fillerData=\[(?<result>.+)\];/,
            )?.groups?.result;
            episodesList = processFillerAndUpdateEpisodeList(
              fillerListRegexMatchResult,
              episodesList,
            );
          });
        const statusText = $('#status')
          .text()
          .match(/Status : (?<status>\w+)/).groups;
        const {status} = statusText;
        const genres = $('#genredata').text().split(', ');
        const animetitle = $('.animetitle').text();
        return {
          episodesList,
          malid,
          status,
          genres,
          animetitle,
        };
      }
    } catch {
      /* empty */
    }
    throw new Error('No response');
  }

  getStreamingLinks({episodeNumber}) {
    const parsedNumber = String(parseInt(episodeNumber, 10) - 1);
    const link = this.#streamingLinksObj?.[parsedNumber]
      ? this.#streamingLinksObj[parsedNumber]?.replace(
          'gogohd.net',
          this.#gohostname,
        )
      : '';
    const linkRegexMatchResult = link.match(
      /streaming\.php\?id=(?<result>.+)&title=/,
    )?.groups?.result;
    let internalPlayerLink;
    if (linkRegexMatchResult) {
      internalPlayerLink = `${getAnimixplayBaseUrl()}/api/cW9${btoa(
        `${linkRegexMatchResult}LTXs3GrU8we9O${btoa(linkRegexMatchResult)}`,
      )}`;
    }
    return {externalPlayerLink: link, internalPlayerLink};
  }
}

export default Episode;
