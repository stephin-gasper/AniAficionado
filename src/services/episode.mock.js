export const EPISODE_DATA = {
  animetitle: 'Boruto: Naruto Next Generations',
  episodesList: [
    {
      isFiller: false,
    },
    {
      isFiller: true,
    },
    {
      isFiller: true,
    },
    {
      isFiller: true,
    },
    {
      isFiller: true,
    },
    {
      isFiller: false,
    },
    {
      isFiller: false,
    },
    {
      isFiller: false,
    },
    {
      isFiller: true,
    },
    {
      isFiller: false,
    },
    {
      isFiller: false,
    },
    {
      isFiller: true,
    },
    {
      isFiller: true,
    },
    {
      isFiller: true,
    },
    {
      isFiller: false,
    },
  ],
  genres: ['Action', 'Adventure', 'Fantasy', 'Martial Arts', 'Shounen'],
  malid: '34566',
  status: 'Ongoing',
};

export const STREAMING_LINKS = {
  externalPlayerLink:
    '//gogohd.net/streaming.php?id=OTc2MzI=&title=Boruto%3A+Naruto+Next+Generations+Episode+1',
  internalPlayerLink:
    'https://animixplay.to/api/cW9T1RjMk16ST1MVFhzM0dyVTh3ZTlPVDFSak1rMTZTVDA9',
};
