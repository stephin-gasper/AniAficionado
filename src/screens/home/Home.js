import React, {useCallback, useState} from 'react';
import {View, InteractionManager} from 'react-native';
import {getLatestEpisodes} from 'services/anime';
import {useFocusEffect} from '@react-navigation/native';

import Cards from './cards';

const Home = () => {
  const [latestEpisodes, setLatestEpisodes] = useState([]);

  useFocusEffect(
    useCallback(() => {
      const task = InteractionManager.runAfterInteractions(() => {
        getLatestEpisodes().then((data) => {
          setLatestEpisodes(data);
        });
      });

      return () => task.cancel();
    }, []),
  );

  return (
    <View>
      {latestEpisodes.map((episode) => (
        <Cards key={episode.id} {...episode} />
      ))}
    </View>
  );
};

export default Home;
