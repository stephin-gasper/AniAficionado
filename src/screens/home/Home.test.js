import {View as MockView, InteractionManager} from 'react-native';
import React from 'react';
import {render, act} from '@testing-library/react-native';
import {getLatestEpisodes} from 'services/anime';

import Home from './Home';

const mockUseFocusEffect = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useFocusEffect: (params) => mockUseFocusEffect(params),
}));

jest.mock('services/anime');

jest.mock('./cards', () => (props) => (
  <MockView testID="animeCard" {...props} />
));

describe('<Home />', () => {
  it('should call api to get latest anime episodes', async () => {
    const {getAllByTestId} = render(<Home />);
    await act(async () => {
      mockUseFocusEffect.mock.calls[0][0]();
    });
    expect(getAllByTestId('animeCard')).toHaveLength(2);
    expect(getLatestEpisodes).toHaveBeenCalledTimes(1);
    expect(getLatestEpisodes).toHaveBeenCalledWith();
  });

  it('should render correctly', async () => {
    const container = render(<Home />);

    await act(async () => {
      mockUseFocusEffect.mock.calls[0][0]();
    });

    expect(container).toMatchSnapshot();
  });

  it('should cancel interaction manager on unmount', async () => {
    render(<Home />);
    const mockCancel = jest.fn();
    jest
      .spyOn(InteractionManager, 'runAfterInteractions')
      .mockReturnValue({cancel: () => mockCancel()});
    let useFocusEffectCallbackInstance = jest.fn();
    await act(async () => {
      useFocusEffectCallbackInstance = mockUseFocusEffect.mock.calls[0][0]();
    });
    await act(async () => {
      useFocusEffectCallbackInstance();
    });

    expect(mockCancel).toHaveBeenCalledTimes(1);
  });
});
