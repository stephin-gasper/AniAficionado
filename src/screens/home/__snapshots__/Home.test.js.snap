// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render correctly 1`] = `
<View>
  <View
    id={4667}
    imageUrl="https://i.animepahe.com/snapshots/kenja-no-deshi-wo-nanoru-kenja.jpg"
    latestEpisodeNumber={9}
    testID="animeCard"
    title="Kenja no Deshi wo Nanoru Kenja"
  />
  <View
    id={4664}
    imageUrl="https://i.animepahe.com/snapshots/the-genius-princes-guide-to-raising-a-nation-out-of-debt.jpg"
    latestEpisodeNumber={9}
    testID="animeCard"
    title="Tensai Ouji no Akaji Kokka Saisei Jutsu"
  />
</View>
`;
